p2p: Track orphans by who provided them

ajtowns opened the pull request:
We currently process orphans by assigning them to the peer that provided a missing parent; instead assign them to the peer that provided the orphan in the first place. This prevents a peer from being able to marginally delay another peer's transactions and also simplifies the internal API slightly. Because we're now associating orphan processing with the peer that provided the orphan originally, we no longer process orphans immediately after receiving the parent, but defer until a future call to `ProcessMessage`.

Based on #26295

ajtowns committed:
txorphanage: index workset by originating peer

ajtowns committed:
net_processing: only process orphans before messages

Previously, when we processed a new tx we would attempt to ATMP any
orphans that considered the new tx a parent immediately, but would only
accept at most one such tx, leaving any others to be considered on a
future run of ProcessMessages(). With this patch, we don't attempt any
orphan processing immediately after receiving a tx, instead deferring
all of them until the next call to ProcessMessages().

ajtowns committed:
txorphange: Drop redundant originator arg from GetTxToReconsider

glozow reviewed (commented):
Concept ACK

ajtowns commented:
Some previous discussion at https://github.com/bitcoin/bitcoin/pull/21527#discussion_r601221165

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2889,7 +2889,7 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
 bool PeerManagerImpl::ProcessOrphanTx(Peer& peer)
 {
     AssertLockHeld(g_msgproc_mutex);
-    AssertLockHeld(cs_main);
```
Should become AssertLockNotHeld? :)

ajtowns replied:
Seems more reasonable to continue to allow invoking `ProcessOrphanTx` with `cs_main` held, even though we could forbid it. (Before the "only process orphans before messages" commit, `ProcessOrphanTx()` was called from `ProcessMessage()` with `cs_main` held)

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -585,10 +585,10 @@ class PeerManagerImpl final : public PeerManager
      * @peer[in]  peer     The peer whose orphan transactions we will reconsider. Generally only one
-     *                     orphan will be reconsidered on each call of this function. This set
-     *                     may be added to if accepting an orphan causes its children to be
+     *                     orphan will be reconsidered on each call of this function. May
```
1. I think it should say "Creates more work if", as it's less ambiguous (I actually spent time figuring the difference out...).
2. So now `bool fMoreNodeWork = m_msgproc->ProcessMessages(pnode, flagInterruptMsgProc);` should be renamed to `may_have_more_work` right?

ajtowns replied:
I don't think this PR changes whether `fMoreNodeWork` should be renamed -- the bool should have the same properties both before and after this PR.

But... I think its current behaviour is technically not quite right? If we process a TX message in `ProcessMessage` now, we'll set `fMoreWork = true` only when `m_getdata_requests` has entries, but I think we should also be doing it if the peer's orphanage work queue isn't empty, as `ProcessMessage` may have added to it. I think this is technically a bug introduced in #15644 -- there should have been an `if (!pfrom->orphan_work_set.empty()) fMoreWork = true;`  added after the `ProcessMessage()` call, as well as after the `// this maintains the order of responses` comment.

The only result of the bug is that when a parent comes in, processing its orphaned children will be delayed by 100ms rather than occurring essentially immediately.

ajtowns replied:
Maybe it would be better to write:

```c++
while (true) {
    bool all_nodes_finished_processing = true;
    for (pnode : nodes) {
        if (m_msgproc->ProcessMessages(pnode)) all_nodes_finished_processing = false;
        m_msgproc->SendMessages(pnode);
    }
    if (all_nodes_finished_processing) wait_for(100ms);
}
```

The idea (as I understand it) is just that `ProcessMessages()` returns true if it does some substantial work and exits early, without checking everything. If that's the case for any peer, then it doesn't wait before looping through all peers again. If `ProcessMessages()` completed for all peers, whether it did work or not, then it is okay to sleep briefly before looping over everything again.

naumenkogs replied:
I think you are right. Let's try applying these changes and see whether everything makes sense once again.

Part of my confusion is that it's unclear whether `has_more_work` applies to a given peer, or to all peers... That variable rename could help. (`peer_has_more_work` vs `node_has_more_work` or something.)

naumenkogs commented:
Concept ACK

ajtowns committed:
net_processing: indicate more work to do when orphans are ready to reconsider

When PR#15644 made orphan processing interruptible, it also introduced a
potential 100ms delay between processing of the first and second newly
reconsiderable orphan, because it didn't check if the orphan work set
was non-empty after invoking ProcessMessage(). This adds that check, so
that ProcessMessages() will return true if there are orphans to process,
usually avoiding the 100ms delay in CConnman::ThreadMessageHandler().

ajtowns commented on src/net_processing.cpp:
```diff
@@ -4907,6 +4907,7 @@ bool PeerManagerImpl::ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt
             LOCK(peer->m_getdata_requests_mutex);
             if (!peer->m_getdata_requests.empty()) fMoreWork = true;
         }
+        if (m_orphanage.HaveTxToReconsider(peer->m_id)) fMoreWork = true;
```
This doesn't work after the other changes in this PR -- you might be creating more work for some other peer, so should be checking if any peers have txs to reconsider, and only doing this once after looping through all the peers.

ajtowns replied:
So, I guess this raises the question: should we move orphan processing outside of `ProcessMessages` and `SendMessages` entirely, and just do it separately, independent of both who sent the orphan and who sent the parent?

Unless someone has a good reason not to, I think I might draft that up as an alternative PR so the approaches can be compared more directly.

glozow replied:
> should we move orphan processing outside of ProcessMessages and SendMessages entirely, and just do it separately, independent of both who sent the orphan and who sent the parent?

Had a similar thought, mainly I don't see why we should try to process all orphans before moving on to `ProcessMessage()`... afaict we never don't a message to a peer as a result of processing an orphan (other than potentially disconnecting) so maintaining the order of responses is not justification like it is for finishing `ProcessGetData()` queue.

fwiw doing orphan processing independently seems reasonable to me, though I don't have a very strong grasp of all the possible implications.

ajtowns replied:
I thought of it more along the lines of "if a node gives us a lot of orphans to process, and we're now processing them, we shouldn't do more work for that node until all the orphan work is done; also we shouldn't delay work from other nodes while we're processing that node's orphans".

mzumsande replied:
> So, I guess this raises the question: should we move orphan processing outside of `ProcessMessages` and `SendMessages` entirely,

Where do you intend to move it? Into a dedicated thread for orphan processing?

> This doesn't work after the other changes in this PR -- you might be creating more work for some other peer, so should be checking if any peers have txs to reconsider, and only doing this once after looping through all the peers.

Is this situation (that the peer who sent us an orphan is different from the peer that sent us the parent) typical? Otherwise might it be acceptable to just have a unnecessary 100ms wait if it occurs very rarely?

ajtowns replied:
> Where do you intend to move it? Into a dedicated thread for orphan processing?

No, I was thinking a separate `ProcessOrphans` method like `ProcessMessages` and `SendMessages`.

> Is this situation (that the peer who sent us an orphan is different from the peer that sent us the parent) typical? Otherwise might it be acceptable to just have a unnecessary 100ms wait if it occurs very rarely?

On a long running node, I'm seeing maybe a few hundred orphans being accepted during startup (ie, children whose parent txs were announced while my node was offline) but then perhaps only 20-60 accepted orphans per day. I don't have much of an idea how many of those would have a different node sending the parent compared to the child, but it doesn't seem likely to be a real problem.

jnewbery replied:
> should we move orphan processing outside of ProcessMessages and SendMessages entirely, and just do it separately, independent of both who sent the orphan and who sent the parent?
> 
> Unless someone has a good reason not to, I think I might draft that up as an alternative PR so the approaches can be compared more directly.

See #19364

mzumsande reviewed (commented):
Concept ACK

ajtowns commented:
Updated. I left the orphan processing in `ProcessMessage` based on the following thoughts:

 * an attacker could probably create arbitrary amounts of work by sending orphan txs that then gets added to the todo list as soon as the parent becomes available. if they do this, it should only slow down communication with that peer, not any of our other peers.
 * avoiding a pretty rare 100ms delay isn't worth the hassle of extending `NetEventsInterface` with new virtual methods for handling orphan txs

naumenkogs commented on src/txorphanage.h:
```diff
@@ -29,11 +29,11 @@ class TxOrphanage {
      *  to work on. Otherwise returns the transaction reference, removes
-     *  the transaction from the work set, and populates its arguments with
-     *  the originating peer, and whether there are more orphans for this peer
+     *  the transaction from the work set, and populates "more" with
```
nit: does "with and whether" make any sense? Not for me, so might be a typo.

ajtowns replied:
fixed

naumenkogs commented:
utACK 6c0a068ea3670bb7a1c955ed6ecde5d050c1a812

ajtowns committed:
net_processing: Don't process tx after processing orphans

If we made progress on orphans, consider that enough work for this peer
for this round of ProcessMessages. This also allows cleaning up the api
for TxOrphange:GetTxToReconsider().

naumenkogs commented:
utACK c58c249a5b694c88122589fedbef4e2f13f08bb4

glozow commented on src/net_processing.cpp:
```diff
@@ -4902,6 +4902,12 @@ bool PeerManagerImpl::ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt
             LOCK(peer->m_getdata_requests_mutex);
             if (!peer->m_getdata_requests.empty()) fMoreWork = true;
         }
+        // Does this peer has an orphan ready to reconsider?
```
c58c249a5b694c88122589fedbef4e2f13f08bb4 nit: s/has/have

glozow reviewed (commented):
ACK c58c249a5b694c88122589fedbef4e2f13f08bb4

mzumsande commented on src/txorphanage.h:
```diff
@@ -48,6 +48,9 @@ class TxOrphanage {
     void AddChildrenToWorkSet(const CTransaction& tx) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);;
 
+    /** Does this peer have any work to do? */
+    bool HaveTxToReconsider(NodeId peer) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);;
```
nit: double `;`

mzumsande reviewed (commented):
Code Review ACK c58c249a5b694c88122589fedbef4e2f13f08bb4

nits can be ignored.

Wenyu1227 reviewed (commented):
src/net_processing.cpp
